CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/clean/SplitNodesPlugin.java

```diff
@@ -130,82 +144,132 @@ public void updateParameters(Graph graph, PluginParameters parameters) {
+        int vertexAttributeCount = graph.getAttributeCount(GraphElementType.VERTEX);
 
-        result.add();
-        result.set(GraphRecordStoreUtilities.SOURCE + GraphRecordStoreUtilities.ID, query.get(SOURCE_ID));
-        result.set(GraphRecordStoreUtilities.SOURCE + VisualConcept.VertexAttribute.IDENTIFIER, left);
-        if (ordered_types.size() > 0 && leftVertexTypesMatches.size() > 0) {
-            result.set(GraphRecordStoreUtilities.SOURCE + AnalyticConcept.VertexAttribute.TYPE, leftVertexTypesMatches.get(0));
-        }
-        result.set(GraphRecordStoreUtilities.DESTINATION + VisualConcept.VertexAttribute.IDENTIFIER, right);
+        // Add new node
+        int newVertexId = graph.addVertex();
 
         //Loops through all of the Node attributes and copies them to the new node
-        for (final String key : query.keys()) {
-            if (key.endsWith(".[id]") || SOURCE_IDENTIFIER.equals(key)) {
-                //Skips the id and Identifier to make the new node unique
-            } else if ((GraphRecordStoreUtilities.SOURCE + VisualConcept.VertexAttribute.X).equals(key)
-                    || (GraphRecordStoreUtilities.SOURCE + VisualConcept.VertexAttribute.Y).equals(key)
-                    || (GraphRecordStoreUtilities.SOURCE + VisualConcept.VertexAttribute.Z).equals(key)) {
-                //The coordinates are also skipped so that the second node is not created in the exact same location
-                //as the first node
+        for (int vertexAttributePosition = 0; vertexAttributePosition < vertexAttributeCount; vertexAttributePosition++) {
+            int vertexAttributeId = graph.getAttribute(GraphElementType.VERTEX, vertexAttributePosition);
+            String vertexAttributeName = graph.getAttributeName(vertexAttributeId);
+
+            if (vertexAttributeName.equals(VisualConcept.VertexAttribute.IDENTIFIER.getName())) {
+                graph.setStringValue(vertexIdentifierAttributeId, newVertexId, newNodeIdentifier);
+            } else if ((VisualConcept.VertexAttribute.X.getName()).equals(vertexAttributeName)) {
+                //The X coordinate is skipped so that the second node is not created in the exact same location
+                //as the first node. Copiying the Y & Z for a better arrangement as it looks less cluttered.
             } else {
-                result.set(GraphRecordStoreUtilities.DESTINATION + key.replace(GraphRecordStoreUtilities.SOURCE, ""), query.get(key));
+                graph.getNativeAttributeType(vertexAttributeId).copyAttributeValue(graph, vertexAttributeId, selectedNode, newVertexId);
             }
         }
 
-        if (ordered_types.size() > 1 && rightVertexTypesMatches.size() > 0) {
-            result.set(GraphRecordStoreUtilities.DESTINATION + AnalyticConcept.VertexAttribute.TYPE, rightVertexTypesMatches.get(0));
+        // Transactions
+        final boolean directed = false;
```

Comment:
if directed can't ever be true, it's probably better to remove this variable and just simply add false into the relevant function calls

Reply:
Whatever the value added here will be overwritten when copying the attributes. Just wanted a default value here. I vaguely remember Sanarqube throwing a code smell  when I added "false" like value in few lines without using a variable.

Reply:
@Auriga2 Ok I think I understand what you are doing here.

So what you have done works, but it's not optimal. The directed parameter of addTransaction() is actually the direction under the hood (e.g. Graph.UPHILL, Graph.UNDIRECTED) and not the value of the directed transaction attribute. Copying the value of the transaction parameter works because when you complete schema, the function checks for a mismatch in transaction attribute value vs under the hood value (e.g. transaction attribute is true but the under the hood value is Graph.UNDIRECTED) and creates a copy of the transaction with the correct under the hood value (removing the old one) if a mismatch exists. This means that with what you currently have, any directed transactions you are copying are effectively being created twice (once by your function and once on complete schema).

I would suggest then grabbing the value of the directed attribute beforehand and passing that value in to addTransaction() to ensure the transaction only gets added once

Reply:
@antares1470  I don't quite get it. If it's `Graph.UPHILL`, `Graph.UNDIRECTED` etc, and not `true` or `false`, then it's still the same thing when I grab and use the directed attribute (which is` true` or `false`) when creating transaction.

Reply:
![image](https://user-images.githubusercontent.com/64826928/90600655-1d8b4800-e236-11ea-9fa6-0f976c0628bc.png)
 I don't get the relation of this boolean to the `GraphConstants`.

Reply:
@Auriga2 so when you add a transaction, the directed parameter determines the under-the-hood direction. The value of the directed transaction attribute is then determined from this under-the-hood value. The main issue is that this under-the-hood value can't actually be changed on a transaction (and so changing the value of the directed transaction attribute after the fact doesn't actually change this value). Consequently when you do change the directed transaction value, a new transaction needs to created with the correct under-the-hood value. This is the relevant code in `completeTransaction()` in `AnalyticSchema` that does that:
```
final Boolean directed = graph.getBooleanValue(transactionDirectedAttribute, transactionId);
final boolean transactionIsDirected = graph.getTransactionDirection(transactionId) != Graph.FLAT;
if (directed != transactionIsDirected) {
   <code to create clone of transaction>
}
```
What this means is that when adding a transaction, it's better to pass through the right value for the directed parameter rather than leaving it to `completeSchema` to fix it after the fact.

Does that make sense?

Reply:
It makes sense to provide the correct value beforehand, and verified that it does no longer attempt to remove and add the transactions again by the above code.